File: src/clu/sources/system_version_plist.py
import logging
import plistlib

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts, Tier
from clu.sources import Source, PARSE_FAIL_MSG
from clu.input import text_file

log = logging.getLogger(__name__)


class SystemVersionPlist(Source):
    _keys = [
        "os.name",
        "os.version",
        "os.build",
        "id.build_id",
    ]

    def provides(self, provides: Provides) -> None:
        for key in self._keys:
            provides[key] = self

    def requires(self, requires: Requires) -> None:
        requires.files.append("/System/Library/CoreServices/SystemVersion.plist")

    def parse(self, facts: Facts) -> None:
        if "os.name" in facts:
            return

        data = text_file("/System/Library/CoreServices/SystemVersion.plist")
        plist = plistlib.loads(bytes(data, "utf-8"))
        log.debug(f"{plist=}")

        if not plist:
            for key in self._keys:
                facts[key] = PARSE_FAIL_MSG
            return
        else:
            facts.add(Tier.ONE, "os.name", plist.get("ProductName", PARSE_FAIL_MSG))
            facts.add(Tier.ONE, "os.version", plist.get("ProductVersion", PARSE_FAIL_MSG))
            facts.add(Tier.TWO, "os.build", plist.get("ProductBuildVersion", PARSE_FAIL_MSG))
            facts.add(Tier.THREE, "id.build_id", plist.get("BuildID", PARSE_FAIL_MSG))

File: pkg/facts/sources/system_version_plist.go
package sources

import (
	"strings"

	"github.com/NHGRI/clu/pkg/facts/types"
	"github.com/NHGRI/clu/pkg/input"
)

// SystemVersionPlist parses macOS SystemVersion.plist for OS info.
type SystemVersionPlist struct{}

var systemVersionFacts = map[string]*types.Fact{
	"os.name":     {Name: "os.name", Tier: types.TierOne, Origin: "ProductName"},
	"os.version":  {Name: "os.version", Tier: types.TierOne, Origin: "ProductVersion"},
	"os.build":    {Name: "os.build", Tier: types.TierThree, Origin: "ProductBuildVersion"},
	"id.build_id": {Name: "id.build_id", Tier: types.TierThree, Origin: "BuildID"},
}

func (s *SystemVersionPlist) Provides(p types.Provides) {
	for name := range systemVersionFacts {
		p[name] = s
	}
}

func (s *SystemVersionPlist) Requires(r *types.Requires) {
	r.Files = append(r.Files, "/System/Library/CoreServices/SystemVersion.plist")
}

func (s *SystemVersionPlist) Parse(f *types.FactDB) {
	if f.Contains("os.name") {
		return
	}

	path := "/System/Library/CoreServices/SystemVersion.plist"
	data, err := input.FileReader(path)
	if err != nil {
		// failures are tier one to make sure they are visible
		for _, fact := range systemVersionFacts {
			fact.Value = types.ParseFailMsg
			f.AddFact(*fact)
		}
		return
	}

	// Simple extraction for expected keys; not a full plist parser.
	for _, fact := range systemVersionFacts {
		fact.Value = extractTag(data, fact.Origin)
	}

	// Add all facts to the FactDB
	for _, fact := range systemVersionFacts {
		f.AddFact(*fact)
	}
}

func extractTag(s, key string) string {
	// FIXME: This is a very naive xml plist parser; consider using a proper xml  library. (AI did this)
	// Look for <key>key</key><string>value</string>
	keyTag := "<key>" + key + "</key>"
	idx := strings.Index(s, keyTag)
	if idx == -1 {
		return ""
	}
	rest := s[idx+len(keyTag):]
	start := strings.Index(rest, "<string>")
	end := strings.Index(rest, "</string>")
	if start == -1 || end == -1 || end <= start+len("<string>") {
		return ""
	}
	return rest[start+len("<string>") : end]
}
